### DISCUSSION TITLE
Should BeginPaint return a SafeHandle?

### DISCUSSION THREAD

**Discussion by rgwood** [microsoft/win32metadata#582] [Q&A]
I'm implementing some of the classic examples from Petzold's Programming Windows using CsWin32 (as a learning exercise), and I ran into a minor issue converting the following code from [HelloWin.c](https://resources.oreilly.com/examples/9781572319950/blob/cbc15539f2b155a2b282c1f20a23cd2a0a13ed4d/cd_contents/Chap03/HelloWin/HelloWin.c)'s WndProc:

## Original

```c
     case WM_PAINT:
          hdc = BeginPaint (hwnd, &ps) ;
          
          GetClientRect (hwnd, &rect) ;
          
          DrawText (hdc, TEXT ("Hello, Windows 98!"), -1, &rect,
                    DT_SINGLELINE | DT_CENTER | DT_VCENTER) ;
          
          EndPaint (hwnd, &ps) ;
          return 0 ;
```

## My C# code

```csharp
case Constants.WM_PAINT:
    hdc = PInvoke.BeginPaint(hwnd, out ps);
    PInvoke.GetClientRect(hwnd, out rect);

    PInvoke.DrawText(hdc, "Hello World", -1, ref rect, DRAW_TEXT_FORMAT.DT_SINGLELINE |  DRAW_TEXT_FORMAT.DT_CENTER | DRAW_TEXT_FORMAT.DT_VCENTER);
    PInvoke.EndPaint(hwnd, ps);
    break;
```

This fails to compile with `CS1503	Argument 1: cannot convert from 'Windows.Win32.Graphics.Gdi.HDC' to 'System.Runtime.InteropServices.SafeHandle'`. `PInvoke.DrawText()` expects a SafeHandle but `PInvoke.BeginPaint()` does not return a SafeHandle implementation.

I can wrap the HDC in a SafeHandle myself, but I *think* CsWin32 should be doing this for me (the README says "SafeHandle-types automatically generated"). Is that correct or am I missing something here?

(Version info: CsWin32 0.1.506-beta, Visual Studio 16.11.0 Preview 1.0, .NET 6 Preview 4)

**Comment by AArnott** [comment]
This is dictated by the metadata. I don't have time to review the case right now, but I'll move to the repo where the decision and change would need to be made.

**Reply by mikebattista** [reply]
The metadata looks in sync to me. All of these functions return/take an HDC.

Is C#/Win32 projecting them in a way that makes them incompatible?

**Reply by AArnott** [reply]
@rgwood: there *is* an overload of `DrawText` that accepts an `HDC` parameter. So calling that is one option.
It looks like the friendly overload of `BeginPaint` that you're calling should be returning a `ReleaseDCSafeHandle`. I'm not sure why it isn't. I'm investigating.

**Reply by AArnott** [reply]
Oh! It's because `ReleaseDC` is a special case in that it takes *two* parameters. Tracked by #423 in the metadata and https://github.com/microsoft/CsWin32/issues/209 in CsWin32.

**Comment by mikebattista** [comment]
Duplicate of https://github.com/microsoft/CsWin32/issues/209
